components/sup/src/manager/service/spec.rs

```diff
@@ -266,6 +266,113 @@ impl ServiceSpec {
         }
         Ok(self)
     }
+
+    /// Given an `old` and a `new` spec, figure out what operations
+    /// are needed in order to turn the `old` state into the `new`
+    /// state.
+    ///
+    /// Here, `old` represents what a currently running service, while
+    /// `new` represents a new version of that spec that we wish to
+    /// make the currently running service. Both are `Option`s, in
+    /// order to capture the scenario in which, say, nothing is
+    /// currently running, but we wish to start a service, or where we
+    /// are running a service, but wish to stop it.
+    ///
+    /// Currently, it is *assumed* that both specs (when present)
+    /// refer to the same service.
+    ///
+    /// Returning `None` indicates that no operation is required.
+    pub(crate) fn reconcile(old: Option<ServiceSpec>,
+                            new: Option<ServiceSpec>)
+                            -> Option<ServiceOperation> {
+        // TODO (CM): Invariant:
+        // old.is_some() && new.is_some() => old.ident == new.ident
+
+        // We need to compare the old spec to the new spec, taking
+        // into consideration the desired state of each. While we can
+        // do that via pattern matching directly, it gets a little
+        // hairy. Instead, we'll just extract the data we need into
+        // one unified match statement and go from there.
+        use DesiredState::{Down,
+                           Up};
+        match (old.as_ref().map(|o| o.desired_state),
```

Comment:
This is awesome! Trying to express this sort of logic without rust's pattern matching gives me nightmares.

Reply:
This was one of my favorite tricks I picked up from Erlang :smile: